File: nodejs/valid_parentheses.js

// validate parentheses of string

var inputChars = process.argv[2].split('');

var openParens = ['(', '[', '{'];
var closeParens = [')', ']', '}'];
var parenPairs = {
    ')': '(',
    ']': '[',
    '}': '{'
};
var parenHistory = [];
var parensValid = true;

inputChars.forEach(function(char) {
    if(!parensValid) return;
    // open parentheses found
    if(openParens.indexOf(char) > -1) {
        parenHistory.push(char);
    }
    // close parentheses found
    else if(closeParens.indexOf(char) > -1) {
        var lastOpen = parenHistory.pop();
        if(lastOpen === undefined || lastOpen !== parenPairs[char]) {
            parensValid = false;
        }
    }
});

if(parenHistory.length > 0) parensValid = false;

console.log(parensValid);

File: python/valid_parentheses.py

# validate parentheses of string

import sys

inputChars = [ x for x in sys.argv[1] ]

openParens = ('(', '[', '{')
closeParens = (')', ']', '}')
parenPairs = {
    ')': '(',
    ']': '[',
    '}': '{'
}
parenHistory = []

for c in inputChars:
    if c in openParens:
        parenHistory.append(c)
    elif c in closeParens:
        if len(parenHistory) == 0 or parenHistory.pop() != parenPairs[c]:
            print False
            sys.exit(1)

print True if len(parenHistory) == 0 else False

